### ISSUE TITLE
Refactor the Renderer interface to use io.Writer instead of *bytes.Buffer

### ISSUE THREAD

**Issue by lukaslueg** [russross/blackfriday#120]
`markdown.Renderer` currently takes a `*bytes.Buffer` as first argument for all functions. This limits the possibilities to design new `Renderer`s which rely on internal state while parsing is going on. In particular this prevents us from implementing `Renderer`s that do not process markdown into a series of bytes alone but, for example, as I would like to do, use ncurses to render the output.

The obvious fix is to define all those functions as taking any `io.Writer` as first argument. The already written code sees almost no changes except some `out.WriteString`-calls changing to `out.Write(io.WriteString(...` and such.

I could make the changes, if this was something to be considered getting pulled.

**Comment by mholt** [comment]
What about the cases where out.Truncate(...) needs to be called, in the case where text() is a function passed in, and returns false? How would that be done with an io.Writer?

**Comment by rtfb** [comment]
TL;DR: by not calling `out.Truncate()` :-)

Elimination of callbacks could look like this: https://github.com/rtfb/blackfriday/commit/14df2451b293351d730d0bfcc6293cdf5b690cf1

Then calls to `Truncate()` can be avoided by more careful triggering of parse events, like here: https://github.com/rtfb/blackfriday/commit/cfdbd5678e6eef6ac8dbcac8e1a29b2858cf6632

The above commits are from this branch here: https://github.com/rtfb/blackfriday/commits/v2, where I was working on these things sporadically for some time. And it seems to work, in essence, but metric ton of tests still fail.

**Comment by russross** [comment]
I just want to chime in and say that these kinds of cleanups would be great. Blackfriday is pretty ugly code in a lot of ways. It closely follows an old C codebase, which in turn was a close translation of the Perl markdown code, which is why you see a lot of code that is basically an expansion of an old regular expression.

I also wonder if it is worth all the effort to fix those problems. A while back I looked at porting John MacFarlane's CommonMark code over to Go. It parses everything into an AST and then generates the HTML output from that, which is a much nicer structure. It was also easy to structure it as an io.Writer. The parsing code is still complicated (unavoidable for any markdown dialect) but I think John understands markdown parsing better than anyone and his code is a great starting point.

The downside to commonmark is that it is missing tables and a few other popular extensions, but I really like it as a way to move forward. It would be nice to have a good Go implementation of commonmark that focuses on tracking the standard, being robust and fuzz proof, and being fast. Then maybe it could act as an upstream provider for more experimental features that some projects need.

BTW--if anyone is interesting in making a commonmark implementation, I think the JavaScript reference implementation is much easier to follow than the C code. The C implementation has had some optimization work that makes it kind of messy and harder to work with.

**Comment by weingart** [comment]
Any idea of when the CommonMark spec will be v1.0?  It seems that the spec has been evolving for some time, and the "reference Go implementation" seems to have been abandoned until v1.0 of the spec is out...

**Comment by rtfb** [comment]
Oh, they do have a Go implementation in their plans? Haven't looked at the details of CommonMark progress for quite a while.

**Comment by client9** [comment]
This appears to be done in v2 now.   Close?

**Comment by rtfb** [comment]
Yes, I guess it's time.
